File: mobile/src/main/java/com/example/android/uamp/MusicServicePresenter.java
package com.example.android.uamp;


import android.content.Intent;
import android.os.Bundle;
import android.os.RemoteException;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import com.example.android.uamp.playback.PlaybackManager;

import static android.app.Service.START_STICKY;

public class MusicServicePresenter {
    private static final int STOP_DELAY = 30000;
    private CustomMediaService service;
    private final MediaSessionCompat mSession;
    private final PlaybackManager mPlaybackManager;
    private final DelayedStopHandler mDelayedStopHandler;
    private MediaNotificationManager mMediaNotificationManager;

    public MusicServicePresenter(MediaSessionCompat mSession, PlaybackManager mPlaybackManager, DelayedStopHandler mDelayedStopHandler) {
        this.mSession = mSession;
        this.mPlaybackManager = mPlaybackManager;
        this.mDelayedStopHandler = mDelayedStopHandler;
    }

    public void setService(CustomMediaService service) {
        this.service = service;
    }

    public void onCreate() {
        service.setSessionToken(mSession.getSessionToken());
        mSession.setCallback(mPlaybackManager.getMediaSessionCallback());
        mSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS | MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS);
        //TODO mSession.setSessionActivity(pi);
        mSession.setExtras(new Bundle());
        mPlaybackManager.updatePlaybackState(null);
        try {
            mMediaNotificationManager = new MediaNotificationManager((MusicService) service);
        } catch (RemoteException e) {
            throw new IllegalStateException("Could not create a MediaNotificationManager", e);
        }
    }

    public int onStartCommand(Intent startIntent, int startId) {
        if (startIntent != null) {
            String action = startIntent.getAction();
            String command = startIntent.getStringExtra(Constants.CMD_NAME);
            if (Constants.ACTION_CMD.equals(action)) {
                if (Constants.CMD_PAUSE.equals(command)) {
                    mPlaybackManager.handlePauseRequest();
                }
            }
        }
        mDelayedStopHandler.removeCallbacksAndMessages(null);
        mDelayedStopHandler.sendEmptyMessageDelayed(0, STOP_DELAY);
        return START_STICKY;
    }

    public void onDestroy() {
        mPlaybackManager.handleStopRequest(null);
        mMediaNotificationManager.stopNotification();
        mDelayedStopHandler.removeCallbacksAndMessages(null);
        mSession.release();
    }

    public void onPlayBackStart() {
        mSession.setActive(true);
        mDelayedStopHandler.removeCallbacksAndMessages(null);
        service.startService(new Intent(service.getApplicationContext(), MusicService.class));
    }

    public void onPlaybackStop() {
        mSession.setActive(false);
        mDelayedStopHandler.removeCallbacksAndMessages(null);
        mDelayedStopHandler.sendEmptyMessageDelayed(0, STOP_DELAY);
        service.stopForeground(true);
    }

    public void onNotificationRequired() {
        mMediaNotificationManager.startNotification();
    }

    public void onPlaybackStateUpdated(PlaybackStateCompat newState) {
        mSession.setPlaybackState(newState);
    }
}

File: app/src/main/java/com/pietrantuono/podcasts/musicservice/MusicServicePresenter.kt
package com.pietrantuono.podcasts.musicservice


import android.app.Service.START_STICKY
import android.content.Intent
import android.os.Bundle
import android.os.RemoteException
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import player.Constants
import player.CustomMediaService
import player.MediaNotificationManager
import player.OtherActions
import player.playback.PlaybackManager

class MusicServicePresenter(
        private val session: MediaSessionCompat,
        private val playbackManager: PlaybackManager,
        private val delayedStopHandler: DelayedStopHandler,
        private val otherActions: OtherActions
) {
    internal var service: CustomMediaService? = null
    private var mMediaNotificationManager: MediaNotificationManager? = null

    val playlist: List<MediaBrowserCompat.MediaItem>
        get() = playbackManager.playlist

    init {
        this.delayedStopHandler.setPresenter(this)
    }

    fun onCreate() {
        service?.setSessionToken(session.sessionToken)
        session.setCallback(playbackManager)
        session.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)
        session.setExtras(Bundle())
        playbackManager.updatePlaybackState(null)
        try {
            mMediaNotificationManager = MediaNotificationManager(service as MusicService?)
        } catch (e: RemoteException) {
            throw IllegalStateException("Could not create a MediaNotificationManager", e)
        }

    }

    fun onStartCommand(startIntent: Intent?, startId: Int): Int {
        if (startIntent != null) {
            if (Constants.ACTION_CMD == startIntent.action) {
                if (Constants.CMD_PAUSE == startIntent.getStringExtra(Constants.CMD_NAME)) {
                    playbackManager.handlePauseRequest()
                }
            }
        }
        delayedStopHandler.removeCallbacksAndMessagesAndSendEmptyMessage()
        return START_STICKY
    }

    fun onDestroy() {
        playbackManager.handleStopRequest(null)
        mMediaNotificationManager?.stopNotification()
        delayedStopHandler.removeCallbacksAndMessages(null)
        session.release()
    }

    fun onPlayBackStart() {
        session.isActive = true
        delayedStopHandler.removeCallbacksAndMessages(null)
        service?.startService(Intent(service?.applicationContext, com.pietrantuono.podcasts.musicservice.MusicService::class.java))
    }

    fun onPlaybackStop() {
        session.isActive = false
        delayedStopHandler.removeCallbacksAndMessagesAndSendEmptyMessage()
        service?.stopForeground(true)
    }

    fun onNotificationRequired() {
        mMediaNotificationManager?.startNotification()
    }

    fun onPlaybackStateUpdated(newState: PlaybackStateCompat) {
        session.setPlaybackState(newState)
    }

    fun stopSelf() {
        service?.stopSelf()
    }

    fun downloadAndAddToQueue(uri: String) {
        otherActions.downloadAndAddToQueue(uri, service)
    }

}
